Some Features For The Python Wrapper

UZerbinati opened the pull request:
I've added some features to the Python wrapper, in particular:

- Wrapped SetIndex Function;
- Added some functions that allow easy export from Netgen to PETSc DMPlex;
- New name attribute and related function.

StefanBruens commented:
I think this has a much better chance of being merged if you rebase it first.

ChrLackner commented on libsrc/include/nginterface_v2.hpp:
```diff
@@ -149,6 +149,7 @@ namespace netgen
     const string * mat;   // material / boundary label
     NG_ELEMENT_TYPE GetType() const { return type; }
     int GetIndex() const { return index-1; }
+    void SetIndex(int ind) { index=ind; }
```
I think since the GetIndex does an index shift between the netgen code and the accessing one, SetIndex should do the reverse one.

UZerbinati replied:
You are correct indeed.

ChrLackner commented on libsrc/meshing/meshclass.cpp:
```diff
@@ -7302,6 +7318,13 @@ namespace netgen
+  std::string Mesh :: GetMeshName(){
+    return meshname;
+  }
+
```
What is the meshname for?

UZerbinati replied:
Is a parameter used in Firedrake, so I added it also here.

ChrLackner commented on libsrc/meshing/python_mesh.cpp:
```diff
@@ -491,6 +491,14 @@ DLL_HEADER void ExportNetgenMeshing(py::module &m)
+                                     for (int i = 0; i < self.GetNP(); i++)
+                                       li.append (py::int_(int(self[i]-1)));
+                                     return li;
+                                   }))
```
Same as el.points but with shift?

ChrLackner commented on libsrc/meshing/meshclass.cpp:
```diff
@@ -7110,6 +7110,22 @@ namespace netgen
+		}
+	}
+	return IDs;
+  }
```
This can easily be done in python as well, is there a reason to not do it there/ make the interface more complex here?

ChrLackner reviewed (commented):
Thanks for the input, but I think most of this can be done as efficiently in Python and does not need separate interfaces. Can you maybe give examples why this wouldn't be (nicely/efficiently) possible in Python?

UZerbinati commented:
The features I've added are needed in order to convert Netgen mesh into a PETSc DMPlex that Firedrake can handle. [Here](https://github.com/firedrakeproject/firedrake/pull/2500) you can find the pull request where this implementation has been discussed.
You are correct most of the things I've added could have been done on the Python side, and as you can see in the pull request linked above this was how it was done originally. But shifting most of the loop over the elements on the C++ side resulted in an implementation that was performing better. In particular, the same code implemented with the loops on the C++ side was 10 times faster (time to solution) than the same code with the loops left on the Python side. I'm not claiming this is in any manner a significant performance test but it was sufficient to convince me to move the loops over the element on the C++ side. Probably the most efficient way to implement the same ideas would be to use pybind11 `memoryview` feature but I'm not quite sure how to use it in this instance.

ChrLackner commented:
Nice to get Netgen into firedrake. I would suggest an even better way though:

I've registered the element types to be used in numpy here:

https://github.com/NGSolve/netgen/commit/c18a317702569e09f78108e124b8e73623647b73

With this you get a numpy view on the element nodes:
```
mesh.ElementsXd().NumPy()["nodes"]
```
with X = 1,2,3. This gives a number_elements x element_max_points numpy matrix with one based node numbers, zero meaning that the entry is not used, so for a first order trig you would get something like [1,2,3,0,0,0,0,0], since element_max_points is 8 (2nd order quads). It is a view on the internal data so no copying is done here.
From there you can use numpy routines to efficiently transform the data to your desired format (cut out the 0s, -1 for 0 based,...)

I think this is the only functionality really needed to get this performant, I think the GetBCIds is fine to be done in python, and you can just give a default "netgen_mesh" name to firedrake.

UZerbinati commented:
First of all, thank you very much for pointing out this feature of netgen I didn't know.
Now let me try to reply feature by feature,

- **Mesh Name**, it would be very good to have a mesh parameter within Netgen because in Firedrake you can work with multiple meshes and the name tag is used when exporting to VTK in order to identify the different meshes. Of course, it would possible to import the Netgen mesh into Firedrake and once it has been converted to a PETSc DMPlex object then we use a Python function to assign it a name, but in my mind, a much more natural flow would be if the Netgen mesh has a name and this attributed is "inherited" by the DMPlex object.
- **GetBCIDs**, keeping the function on the C++ side seems much faster at least if we copy in Python the same code it is now written in C++. You'll find [here](https://github.com/NGSolve/netgen/files/9569834/bench.txt) an example code where I time the two cases, and the result is the following; Time take using loop on BCIDs in Python is: **9.748511806999886**, while keeping loop on C++ side it takes: **0.04197995800132048**.
- **Indicies 2D and 3D**, I'm familiarizing myself with the new feature you implemented. In the next few days I'll modify my Firedrake wrap to use this feature and benchmark it against the old implementation to see what is faster, but as you mentioned since it's a view on internal memory I don't see any reason why using this new feature should be advantageous! 

Thanks again for your help :)

UZerbinati commented:
Dear @ChrLackner ,
I've also added some functions to do mesh refinement of flagged elements directly in NetGen. I didn't find this already in the Python binding, but I probably didn't know where to look.
Best,
Umberto

UZerbinati commented:
ping @ChrLackner

luzpaz commented:
bump

luzpaz commented:
Why close ?

UZerbinati commented:
@luzpaz Most of this stuff can now be done using ngsPETSc :)
